help: tutorials: QT Tutorial

SimonDeplat opened the pull request:
## Purpose and Motivation

Add a step-by-step, beginner friendly QT Tutorial.

Fixes #3449 

## Types of changes

- Documentation

## To-do list

English isn't my native language, so some correction might be needed. I think it is mostly good, but a review would be appreciated.

- [x] Syntax, grammar and phrasing are correct
- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

capital-G commented on HelpSource/Tutorials/QT/compose_complex_view.schelp:
```diff
@@ -0,0 +1,395 @@
+// Done
+window.front;
+)
+::
```
Missing empty line as EOF

capital-G commented on HelpSource/Tutorials/QT/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+Synth(\instrView)
+::
+
+We'll start with the following code :
```
Maybe it is good to spoil that this will not create any sound yet? At least I tried clicking the button in expectation of a sound :)
```suggestion
We'll start with the following code which is not yet connected to our synth and will therefore not yield any sound yet:
```

SimonDeplat replied:
Manually fixed.

capital-G commented on HelpSource/Tutorials/QT/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+
+You will need to strong::boot the Server::, to strong::load the SynthDef:: we will be using in this tutorial. It is a simple sine wave with a percussive enveloppe. We can control its frequency, its volume and its duration.
+
+code::
+(
+SynthDef(\instrView, { |out = 0, amp = 0.25, freq = 440, dur = 1.0|
```
```suggestion
code::
// make sure the server is booted  
s.boot;  

// add our SynthDef to the server  
(
SynthDef(\instrView, { |out = 0, amp = 0.25, freq = 440, dur = 1.0|
```

capital-G commented on HelpSource/Tutorials/QT/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+We just need to introduce this inside the loop that creates the buttons.
+But instead of asking SC to do this 8 times, it will do it for a collection of frequencies instead.
+
+We can strong::transform a set of midi notes into frequencies and map them to the buttons:: :
```
```suggestion
We can strong::transform a set of MIDI notes into frequencies and map them to the buttons:: :
```

SimonDeplat replied:
Manually fixed.

capital-G commented on HelpSource/Tutorials/QT/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+section:: Adapting the Slider to the correct value (mapping)
+
+But what if we want to control other ranges ?
+We can use code::.linlin:: or code::.linexp:: to map the sliders value, which is between 0 and 1, to a value between 0.1 and 2. We'll use this to new value to control the duration of the percussive enveloppe :
```
```suggestion
We can use code::.linlin:: or code::.linexp:: to map the sliders value, which is between 0 and 1, to a value between 0.1 and 2. We'll use this to new value to control the duration of the percussive envelope :
```

SimonDeplat replied:
Manually fixed.

capital-G commented on HelpSource/Tutorials/QT/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+)
+::
+
+Alternatively, we can use a ControlSpec to do the same thing :
```
```suggestion
Alternatively, we can use a link::Classes/ControlSpec:: to do the same thing:
```

capital-G commented on HelpSource/Tutorials/QT/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+// Inserting the volume text before the volume slider :
+windowLayout.add(volumeText, 1);
+windowLayout.add(volumeSlider, 3);
+// Inserting the duration text before the durationslider,
```
```suggestion
// Inserting the duration text before the duration slider,
```

capital-G commented on HelpSource/Tutorials/QT/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+
+section::Conclusion
+
+This has been a long tutorial, but with those basic concepts, you should now be able to manipulate every QT Class without much difficulties.
```
```suggestion
This has been a long tutorial, but with those basic concepts, you should now be able to manipulate every Qt Class without much difficulties.
```

capital-G commented on HelpSource/Tutorials/QT/compose_complex_view.schelp:
```diff
@@ -0,0 +1,395 @@
+title:: QT : Composing views with basic components
+summary:: General examples about composing views with QT base GUI components.
+categories:: Tutorials>QT
```
```suggestion
title:: Qt : Composing views with basic components
summary:: General examples about composing views with Qt base GUI components.
categories:: Tutorials>Qt
```

capital-G commented on HelpSource/Tutorials/QT/compose_complex_view.schelp:
```diff
@@ -0,0 +1,395 @@
+
+section:: What is this tutorial about?
+
+This tutorial presents strong::different interfaces you might need to construct:: when working with QT, strong::and discuss their particularities::. If you need a more detailed tutorial about the basics of view composition, you should read link::Tutorials/QT/create_simple_instrument_view:: first.
```
```suggestion
This tutorial presents strong::different interfaces you might need to construct:: when working with Qt, strong::and discuss their particularities::. If you need a more detailed tutorial about the basics of view composition, you should read link::Tutorials/QT/create_simple_instrument_view:: first.
```

capital-G commented on HelpSource/Tutorials/QT/compose_complex_view.schelp:
```diff
@@ -0,0 +1,395 @@
+
+This tutorial presents strong::different interfaces you might need to construct:: when working with QT, strong::and discuss their particularities::. If you need a more detailed tutorial about the basics of view composition, you should read link::Tutorials/QT/create_simple_instrument_view:: first.
+
+strong::SuperCollider, via QT, implements different GUI components to allow you to compose complex softwares interfaces.::
```
```suggestion
strong::SuperCollider, via Qt, implements different GUI components to allow you to compose complex softwares interfaces.::
```

capital-G commented on HelpSource/Tutorials/QT/compose_complex_view.schelp:
```diff
@@ -0,0 +1,395 @@
+
+This example is the continuation of the previous one, and explains why strong::UI components shouldn't be directly responsible for the action they ultimately perform::. It also talks briefly about focus and keyboard input response.
+
+One of the strong::common mistake when starting using QT is to use a View's action function as if it was effectively the function responsible for modifying the software::. Even if it works reliably if there is only one View associated with the algorithm, this tends to decrease code readability. When several Views are tied to the same algorithm, this almost always introduces issues, as one can see ...
```
```suggestion
One of the strong::common mistake when starting using Qt is to use a View's action function as if it was effectively the function responsible for modifying the software::. Even if it works reliably if there is only one View associated with the algorithm, this tends to decrease code readability. When several Views are tied to the same algorithm, this almost always introduces issues, as one can see in the previous example.
```

capital-G commented on HelpSource/Tutorials/QT/compose_complex_view.schelp:
```diff
@@ -0,0 +1,395 @@
+
+The window is first split vertically in two : a main view, and a small menu that 'contains' hidden views. This menu is only visible when it is non-empty.
+
+For each of our subviews (in our example, a link::Classes/MultiSliderView:: and a link::Classes/StaticText::), we are creating two link::Classes/Button::s : one that hides the view, and is situated in the same layout as the view it is hidding, and one inside the menu, that displays it back. They control the views visibility, and they also control the visibility of the menu.
```
```suggestion
For each of our subviews (in our example, a link::Classes/MultiSliderView:: and a link::Classes/StaticText::), we are creating two link::Classes/Button::s : one that hides the view, and is situated in the same layout as the view it is hiding, and one inside the menu, that displays it back. They control the views visibility, and they also control the visibility of the menu.
```

capital-G commented on HelpSource/Tutorials/QT/compose_complex_view.schelp:
```diff
@@ -0,0 +1,395 @@
+// Add the 'show' button to the visibility view
+visibilityView.layout.add(textShowButton);
+
+// Agencing the top level
```
Agencing?

capital-G commented on HelpSource/Tutorials/QT/things_you_wanna_know.schelp:
```diff
@@ -0,0 +1,61 @@
+title:: QT : Things you probably want to know
+summary:: Miscellaneous things you probably want to know about QT.
+categories:: Tutorials>QT
```
```suggestion
title:: Qt : Things you probably want to know
summary:: Miscellaneous things you probably want to know about Qt.
categories:: Tutorials>Qt
```

capital-G commented on HelpSource/Tutorials/QT/compose_complex_view.schelp:
```diff
@@ -0,0 +1,395 @@
+visibilityView.visible_(false);
+
+
+// --- Here's the visibilty logic : ---
```
```suggestion
// --- Here's the visibility logic : ---
```

capital-G commented on HelpSource/Tutorials/QT/things_you_wanna_know.schelp:
```diff
@@ -0,0 +1,61 @@
+
+SuperCollider provides some useful classes to automate processes. Some of them are associated to a link::Classes/Clock:: : mainly link::Classes/Routine:: and link::Classes/Task::.
+
+strong::QT rendering system is unavailable within the context of clock associated threads::. In other words, you cannot 'directly' modify a widget within a link::Classes/Routine:: or a link::Classes/Task::.
```
```suggestion
strong::Qt rendering system is unavailable within the context of clock associated threads::. In other words, you cannot 'directly' modify a widget within a link::Classes/Routine:: or a link::Classes/Task::.
```

capital-G commented on HelpSource/Tutorials/QT/things_you_wanna_know.schelp:
```diff
@@ -0,0 +1,61 @@
+
+strong::QT rendering system is unavailable within the context of clock associated threads::. In other words, you cannot 'directly' modify a widget within a link::Classes/Routine:: or a link::Classes/Task::.
+
+Instead, you need to defer the modification by strong::encapsulating it in a deferred function::, that will be executed once SuperCollider resume its main context. Doing otherwise will crash the program :
```
```suggestion
Instead, you need to defer the modification by strong::encapsulating it in a deferred function::, that will be executed once SuperCollider resume its main context, see link::Classes/Function#-defer:: Doing otherwise will crash the program :
```

Think it would be good to link to https://doc.sccode.org/Classes/Function.html#-defer , but I am unsure about the actual syntax in this case, please check, have written it from memory and haven't tested it in actually rendered docs.

capital-G commented on HelpSource/Tutorials/QT/things_you_wanna_know.schelp:
```diff
@@ -0,0 +1,61 @@
+
+section:: QT limitations
+
+This remark mostly concerns debugging and issues understanding. Since strong::QT is a third-party program integrated within SuperCollider::, some of its limitations are unrelated to SuperCollider itself.
```
```suggestion
This remark mostly concerns debugging and understanding issues. Since strong::Qt is a third-party library integrated within SuperCollider::, some of its limitations are unrelated to SuperCollider itself.
```

capital-G commented on HelpSource/Tutorials/QT/things_you_wanna_know.schelp:
```diff
@@ -0,0 +1,61 @@
+
+This remark mostly concerns debugging and issues understanding. Since strong::QT is a third-party program integrated within SuperCollider::, some of its limitations are unrelated to SuperCollider itself.
+
+As an example, QT5 link::Classes/QImage:: has a strict pixel limit and cannot exceed 32767x32767 pixels. This 'issue' being only related to QT, it is documented on QT sources/forums, but not on SuperCollider sources/forums. In other words, working with QT within the context of SuperCollider might also include using QTs documentation directly.
```
```suggestion
As an example, Qt5 link::Classes/QImage:: has a strict pixel limit and cannot exceed 32767x32767 pixels. This 'issue' being only related to Qt, it is documented on Qt sources/forums, but not on SuperCollider sources/forums. In other words, working with Qt within the context of SuperCollider might also include using Qts documentation directly.
```

We are actually on the transition to Qt6 which lifts this restriction, so maybe there can be found another example which still creates a restriction on Qt6?

SimonDeplat replied:
Yes, but I'm not aware of any atm. I think this file might be a good place to list common Qt related issues once they're discovered? Or is github sufficient?

SimonDeplat replied:
Updated. Concerning QPainter, I propose to keep it as is for now, and to see later if we need to refine limitations documentation?

capital-G commented on HelpSource/Tutorials/QT/things_you_wanna_know.schelp:
```diff
@@ -0,0 +1,61 @@
+
+The link::Classes/Pen:: class allows the user to define a custom code::drawFunc:: to a link::Classes/View::.
+
+link::Classes/Pen:: usage is a strong::two-steps process:: : the user strong::first 'describes' the shape:: he wants to draw (using methods such as code::fillRect:: or code::lineTo::), strong::then renders the shape::, using either code::fill::, code::stroke::, or code::fillStroke::.
```
```suggestion
link::Classes/Pen:: usage is a strong::two-steps process:: : the user strong::first 'describes' the shape:: they want to draw (using methods such as code::fillRect:: or code::lineTo::), strong::then renders the shape::, using either code::fill::, code::stroke::, or code::fillStroke::.
```

We should not focus on a male-audience within the docs.

SimonDeplat replied:
Sorry about this. I understand and agree with this principle, but this preoccupation is mostly absent from the social context I'm leaving in, so thinking about it isn't an automatism yet.

capital-G commented on HelpSource/Tutorials/QT/things_you_wanna_know.schelp:
```diff
@@ -0,0 +1,61 @@
+
+On the other hand, calling successively code::lineTo:: then code::stroke:: a million time might be time inefficient as several code::lineTo:: could have been grouped together before being drawn on screen.
+
+The best solution in this case would be to test several configurations to balance RAM/CPU usage in the most efficient way.
```
Maybe link to the discussions where this came up for more detail, e.g. https://github.com/supercollider/supercollider/issues/3449

SimonDeplat replied:
I like the idea of documentation being autonomous, requiring the least references possible. And I think references should point to other documentation texts, and this one is an issue.
Tbh, I don't consider #3449 being a real issue, and I think it could've been closed way earlier. But it still is open, and there's so many issues currently pending that every one we can close might be considered a (small) victory. I can't do much but this I can. That's the only reason why I wrote this paragraph down...

That's my point of view: I don't think linking to this issue is actually helpful in term of knowledge, but I recognize my incompetence about this question. If you think linking this is a good idea, I'll do it.

capital-G reviewed (changes requested):
Thanks for your contribution! Any improvements and work on the docs is always appreciated!

I think this tutorial offers a great introduction to the basic of Qt/GUI while also giving a glimpse at more advanced patterns that are useful in the context (e.g. FunctionList).
I also like your iterative approach which allows to follow a "naive" way which people will face and also how to solve them. Also, you sometimes step out of the current context to explain something in a small and isolated example, which works really well IMO.

I added some remarks, mostly concerning some spelling mistakes. I also spotted that you often use a space in front of a `:`, so `you do it like this :` but it would be actually `like this:`
Also the actual name of the library is `Qt` and not `QT`, see https://en.wikipedia.org/wiki/Qt_(software) . I changed where I found it, but the filenames would also need to be renamed to reflect this.

Overall a really great tutorial and a good addition to the SC docs! You should write more tutorials :)

If the minor issues are fixed and another, more proficient person in Qt/GUI has looked over it, I think we can merge it!

edit: note to myself: we should trigger at least the `pre-commit` workflow also on schelp files.

SimonDeplat commented on HelpSource/Tutorials/QT/compose_complex_view.schelp:
```diff
@@ -0,0 +1,395 @@
+// Done
+window.front;
+)
+::
```
I doubt this is the last time I forget an empty line as EOF :) .

"Things you wanna know" is more an appendix to the tutorial, rather than a chapter. In fact, the reason for it is almost only to be able to definitely close #3449 in a 'proper way'. But this also allows for an official disclaimer against QT related issues (which arise from times to times on the forum, or on issues here), and also leaves a space for anyone to document Qt related stuff (for example, #3755, multi-touch currently unsupported, could be mentioned on this file).

SimonDeplat commented:
Thanks a lot for your help and feedback! I've committed your changes, except for the #3449 linking, see my comment about this.
I've changed every `QT` occurrence for `Qt` (including the folder name), and fixed `this :` to `this:` .

Side question : I see you were able to label the PR as "comp: help". I tried to do it but never found a way, does this require privileges ? If not, where do I click :D ?

capital-G commented on HelpSource/Tutorials/Qt/tuto_intro.schelp:
```diff
@@ -0,0 +1,20 @@
+## 1. link::Tutorials/Qt/create_window::: basics of window creation.
+## 2. link::Tutorials/Qt/create_simple_instrument_view::: interfacing a synth (keyboard and volume slider), and introduction to the layout system.
+## 3. link::Tutorials/Qt/compose_complex_view::: documented examples about composed views: a slider with editable text, a tab system, and a minimize system.
+## A. link::Tutorials/Qt/things_you_wanna_know::: miscellaneaous warnings and hints about Qt usage.
```
```suggestion
## A. link::Tutorials/Qt/things_you_wanna_know::: miscellaneous warnings and hints about Qt usage.
```

capital-G reviewed (changes requested):
There are still some typos in it from the last review. If these are fixed I'll go through it once more in full.

edit: merge note: should be squashed

SimonDeplat commented on HelpSource/Tutorials/Qt/compose_complex_view.schelp:
```diff
@@ -0,0 +1,395 @@
+// Add the 'show' button to the visibility view
+visibilityView.layout.add(textShowButton);
+
+// Agencing the top level
```
This why I really need a review :) . This is 'frenglish', a direct translation of the french 'agencer' verb, that sounded English enough to be  written down ! Joke aside, I'm fixing this.

SimonDeplat commented:
I've fixed the issues you mentioned. Thanks again for the help, there's a lot of spelling/grammar rules mistakes I would not have found by myself.
Once I did this, I went through the whole tutorial again, and tried to find every mistake that we might have still missed.
But I have also rephrased some formulations. Those edits are all within the last commit, so I can undo those modifications easily if needed.

capital-G reviewed (approved):
Thanks, everything looks fine to me and I really enjoy this tutorial.

But before a merge I'd prefer if someone who has some more knowledge of QtCollider than me gives this an additional review.

dyfer commented on HelpSource/Tutorials/Qt/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+var button = Button();
+button.string_("Play");
+
+button.action_(
+	// The function that the button
+	// will trigger begins here...
+	{
+		Synth(\instrView);
+		"ting".postln;
+	}
+	// ...and ends here
+);
```
I think it would be better to avoid space/newline/comment between `(` and `{`
```
button.action_({
	// This is the body of the function that the button will trigger
	Synth(\instrView);
	"ting".postln;
});
```

dyfer commented on HelpSource/Tutorials/Qt/create_window.schelp:
```diff
@@ -0,0 +1,180 @@
+var fullScreenActive = false;
+
+View.globalKeyDownAction = {
+	|view, char, mod, unicode, keycode, key|
```
I think the arguments should be in the same line as the opening `{`

dyfer commented on HelpSource/Tutorials/Qt/create_window.schelp:
```diff
@@ -0,0 +1,180 @@
+View.globalKeyDownAction = FunctionList();
+
+View.globalKeyDownAction.addFunc({
+	|view, char, mod, unicode, keycode, key|
```
as above, I think the arguments should be in the same line as the opening `{`

dyfer reviewed (changes requested):
Thanks for this! I'd love to go through the full tutorial before we merge this.

But first - I see that many of the comments from @capital-G were marked as "resolved", while they have not been implemented, unless I'm missing something? If that's the case, please unresolve the rest and implement these changes.

I also submitted a few formatting change requests myself. 
Thanks!

dyfer commented:
@telephon do you think we should pull this in without applying the previously suggested changes?

dyfer commented:
> I implemented @capital-G's requested changes by hand because I thought typos fix should be contained within a single commit.

I _thought_ I saw some changes that were not applied, I'm sorry for the noise. 
I think that some minor things still slipped through, e.g. [non-capitalized word "midi" here](https://github.com/supercollider/supercollider/pull/6451/files#diff-7674b55b4dc3e852c9be6e381e4992aac36d642f05efd51db69b3b220e7a4dfbR175). I'm not going to go through all of the changes though.

> You can see those comments are marked as outdated, meaning they're either fixed by later commits, or phrases have been rewritten entirely.

BTW "Oudated" only means that the file has changed since the comment was made; not necessarily that the changes were made in accordance with the suggestion.

--------

I read through the tutorial - this is a great resource, thanks for writing this.

I have a few minor requests:
- what do you think about adding numbers to file names, so that one could see the order of the tutorial when looking at the help index? (This is just a suggestion, feel free to keep it as is if preferred)

<img width="649" alt="Screenshot 2024-10-21 at 10 57 48 AM" src="https://github.com/user-attachments/assets/58a2d7f2-f50e-49f6-905d-c9ab2ff08cb8">

As an example, see "getting started" tutorial.

- I personally think that "tuto_intro.schelp" is an unnecessary abbreviation that's not really clear. Could you change it to "tutorial_intro.schelp"?
- could you add links to the previous/next section (or at least the next section) under the "see also" tag on top of all of the tutorial pages?
- What do you think about also mentioning `model-view-controller` paradigm in [Centralizing Views actions with a single Function](https://github.com/supercollider/supercollider/pull/6451/files#diff-062fcb8caef04e29f82fbd34d1546f5901ebee7081e86855a97a9e598446c0cfR69) section, also including a link to SimpleController.schelp? I _thought_ that we had a separate guide on MVC paradigm, but I guess not... still, that's another way to implement what you're describing in this section and I think it would be good to point it out.

SimonDeplat commented on HelpSource/Tutorials/QT/create_window.schelp:
```diff
@@ -0,0 +1,180 @@
+
+Section:: And that's it!
+
+With the ability to strong::create a window::, strong::insert views inside:: and make it strong::respond to keyboard inputs::, you're set to go forward in your SuperCollider graphical experimentations.
```
Manually fixed.

SimonDeplat commented on HelpSource/Tutorials/QT/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+section:: Setup
+
+
+You will need to strong::boot the Server::, to strong::load the SynthDef:: we will be using in this tutorial. It is a simple sine wave with a percussive enveloppe. We can control its frequency, its volume and its duration.
```
Manually fixed.

SimonDeplat commented on HelpSource/Tutorials/QT/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+
+section:: What is this tutorial about?
+
+This is a tutorial to get started using SuperCollider's QT implementation to strong::create views and make them interact with a Synth::.
```
Manually fixed.

SimonDeplat commented on HelpSource/Tutorials/QT/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+
+This is a tutorial to get started using SuperCollider's QT implementation to strong::create views and make them interact with a Synth::.
+
+It covers those topics :
```
Manually fixed.

SimonDeplat commented on HelpSource/Tutorials/Qt/create_simple_instrument_view.schelp:
```diff
@@ -0,0 +1,548 @@
+
+You will need to strong::boot the Server::, to strong::load the SynthDef:: we will be using in this tutorial. It is a simple sine wave with a percussive enveloppe. We can control its frequency, its volume and its duration.
+
+code::
+(
+SynthDef(\instrView, { |out = 0, amp = 0.25, freq = 440, dur = 1.0|
```
Manually fixed.

JordanHendersonMusic commented:
Not wanting to interfere too much...

but perhaps (at least in the main headings and search functionality) the term 'Qt' could be removed in favour of 'Graphic User Interface'? 

A new user probably doesn't know what Qt is so they might not find this excellent help file if they were simply looking for how to make a button, which would be a shame!

SimonDeplat commented:
@dyfer Thanks for pointing this out, I definitely missed some of @capital-G changes suggestions. I should have committed them using GitHub's interface instead of out-dating them, reducing commits ain't definitely worth the trouble. Now I know. I went through every requested change again to check if they were fixed. Maybe I still missed one or two, but at least I did my best :) .

---

Now for the indexing - links to the previous/next section. Some of the files have a clear temporal continuation: '_window_' > '_synth view_'. '_composing views_', however, isn't really the direct continuation of '_synth view_'. '_things you might want to know_' is a bit strange because some concepts it covers are good to read _before_ everything else, but some only make sense _after_ reading everything else...

The '_tutorial introduction_' page is somehow useless: the only pertinent information it gives is a recommended tutorial order (so here it is). I wanted to keep the 'unordered' wiki style in the help file system itself, while providing a suggested order in a custom page.

In the future, there's more tutorials to append: UserView's drawFunc, actions, low coupling signal system, etc. Some are related  to others, some are autonomous. Question is: what is SC's help file system philosophy? How do we structure topics and tutorials so that users (and most importantly beginners) feel guided? (also, see [this comment](https://github.com/supercollider/supercollider/issues/771#issuecomment-2393129652))

---

Regarding the `MVC` paradigm, this is a good point, and I'll have a look, but unlike most of the people I can see working on SC, I do not have any technical background, I never studied or worked with computer science. So I've heard about `MVC`, but I can't tell you anything true or intelligent about it, so I'd rather not talk about it for now :) .

---

@JordanHendersonMusic Replacing 'Qt' with something like 'GUI system' is indeed a good idea. I personally dislike abstractions when reading documentation, but I think it would increase accessibility. I'll work on it.

dyfer commented:
> I went through every requested change again to check if they were fixed. 

Thanks so much!

> 
> Now for the indexing - links to the previous/next section. Some of the files have a clear temporal continuation: '_window_' > '_synth view_'. '_composing views_', however, isn't really the direct continuation of '_synth view_'. '_things you might want to know_' is a bit strange because some concepts it covers are good to read _before_ everything else, but some only make sense _after_ reading everything else...
 
> The '_tutorial introduction_' page is somehow useless: the only pertinent information it gives is a recommended tutorial order (so here it is). 

Got it. Yes, some of the information makes sense sequentially. Since you did point out the order in the introduction, I thought the continuation links are stronger, but maybe not.

I still think that adding links under "see also" tag would be helpful - _where applicable_. I think it'd be nice to have the link to the tutorial introduction on every page, and where you indicate there's a stronger connection between them (like `window` -> `synth view`).

> 
> In the future, there's more tutorials to append: UserView's drawFunc, actions, low coupling signal system, etc. Some are related to others, some are autonomous. Question is: what is SC's help file system philosophy? How do we structure topics and tutorials so that users (and most importantly beginners) feel guided? (also, see [this comment](https://github.com/supercollider/supercollider/issues/771#issuecomment-2393129652))


This is a great question. I think this tutorial can go in as is, and then we can discuss the structuring of the information separately.

> 
> Regarding the `MVC` paradigm, this is a good point, and I'll have a look, but unlike most of the people I can see working on SC, I do not have any technical background, I never studied or worked with computer science. So I've heard about `MVC`, but I can't tell you anything true or intelligent about it, so I'd rather not talk about it for now :) .

FWIW I've learned about the mvc paradigm from SC, this wasn't part of my education. But I agree it's fine to leave it out. I think "introduction from MVC" is an idea for another small guide anyway (which then could be linked to your guide and elsewhere).

> @JordanHendersonMusic Replacing 'Qt' with something like 'GUI system' is indeed a good idea. I personally dislike abstractions when reading documentation, but I think it would increase accessibility. I'll work on it.

Are you planning to make this change still?

dyfer commented on HelpSource/Tutorials/GUI/compose_complex_view.schelp:
```diff
@@ -0,0 +1,406 @@
+::
+
+NOTE::
+There are several other ways to interface algorithms with graphical user interfaces. One of the most widely spread is the link::https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller##Model-View-Controller design pattern::.
```
I'd suggest adding the following to this paragraph.

> For SuperCollider implementation of the Model-View-Controller paradigm, see link::Classes/SimpleController::.

dyfer reviewed (changes requested):
Thanks! I think this is really close.

Aside from the SimpleController link, I wondering if there's a good reason to keep [compose_complex_view.schelp](https://github.com/supercollider/supercollider/pull/6451/files#diff-a5524342aa467d681530c96ff0aa46d8dd923585ccd8b3f9cebc3f62cbea7b2b) as a "Qt" tutorial - I think it would make sense to keep with it wiht "GUI" tutorials.

SimonDeplat commented:
This was a mistake. This has been fixed.

Reference to SimpleController has been added.

dyfer reviewed (approved):
I think this is now good to go, thanks so much!

capital-G commented:
Thanks @SimonDeplat - this is a really great resource you contributed! Much appreciated!
